Error: unknown JSON.parse problem when firebase storage deploy

<!-- DO NOT DELETE
validate_template=true
template_path=.github/ISSUE_TEMPLATE/bug_report.md
-->

<!--
Thank you for contributing to the Firebase community!

Think you found a bug?
=======================
Yeah, we're definitely not perfect! Please use this template and include a minimal repro when opening the issue. If you know how to solve the issue, please create a Pull Request, and we'd be happy to review it!

Have a feature request?
========================
Great, we love hearing how we can improve our products! However, do not use this template to submit a feature request. Please submit your feature requests to: https://firebase.google.com/support/contact/bugs-features/

Have a usage question?
=======================
We get lots of those and we love helping you, but GitHub is not the best place for them and they will be closed. Please take a look at the guide first: https://firebase.google.com/docs/cli/

If the official documentation doesn't help, try asking through our official support channel: https://firebase.google.com/support/

Additional locations to check for solutions or assistance from the community:
- Stack Overflow: https://stackoverflow.com/
- Firebase Slack Community: https://firebase.community/

*Please avoid duplicate posting across multiple channels!*
-->

### [REQUIRED] Environment info

<!-- What version of the Firebase CLI (`firebase-tools`) are you using? Note that your issue may already be fixed in the latest versions. The latest version can be found at https://github.com/firebase/firebase-tools/releases -->

<!-- Output of `firebase --version` -->

**firebase-tools:v10.1.3

<!-- e.g. macOS, Windows, Ubuntu -->

**Platform:Github Actions, macOS local

### [REQUIRED] Test case

<!-- Provide a minimal, complete, and verifiable example (http://stackoverflow.com/help/mcve) -->
- do `firebase deploy --force --only storage --project ${{ env.PROJECT_ID }}`

### [REQUIRED] Steps to reproduce

<!-- Provide the steps needed to reproduce the issue with the above test case. -->
- do `firebase deploy --force --only storage --project ${{ env.PROJECT_ID }}`
- its occurred in local command and Github workflow.
-  And it was not happend before the Latest `firebase-tools` update for v10.1.3
### [REQUIRED] Expected behavior

<!-- What is the expected behavior? -->
- firebase storage deploy success

### [REQUIRED] Actual behavior

<!-- Run the command with --debug flag, and include the logs below. -->
- it shows JSON.parse ERROR.
```
i  deploying storage, firestore


There was an issue deploying your functions. Verify that your project has a Google App Engine instance setup at https://console.cloud.google.com/appengine and try again. If this issue persists, please contact support.

Error: Unable to parse JSON: SyntaxError: Unexpected token < in JSON at position 0
Error: Process completed with exit code 1.
```
- This error message seems to be derived from the code [here](https://github.com/firebase/firebase-tools/compare/v10.1.2...v10.1.3#diff-041ee7faf505a1f34aaf8d58f00ba9ccad426e4a2dcc9b84a0a106ff8e2874b7R411-R413)
- Maybe the content of `text` has become `<anonymous>` or `<undefined>`?.

--- a/.github/workflows/node-test.yml
+++ b/.github/workflows/node-test.yml
@@ -73,6 +73,7 @@ jobs:
           - npm run test:emulator
           - npm run test:extensions-emulator
           - npm run test:triggers-end-to-end
+          - npm run test:storage-deploy
           - npm run test:storage-emulator-integration
     steps:
       - uses: actions/checkout@v2

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,1 +1,2 @@
 - Updates the Storage Emulator to use the actual default storage bucket.
+- Fixes issue where `deploy` would fail with a `JSON: SyntaxError` error. (#4117)

--- a/package.json
+++ b/package.json
@@ -30,6 +30,7 @@
     "test:extensions-emulator": "./scripts/extensions-emulator-tests/run.sh",
     "test:hosting": "./scripts/hosting-tests/run.sh",
     "test:triggers-end-to-end": "./scripts/triggers-end-to-end-tests/run.sh",
+    "test:storage-deploy": "./scripts/storage-deploy-tests/run.sh",
     "test:storage-emulator-integration": "./scripts/storage-emulator-integration/run.sh"
   },
   "files": [

--- a/scripts/storage-deploy-tests/run.sh
+++ b/scripts/storage-deploy-tests/run.sh
@@ -40,7 +40,7 @@ rules_version = '2';
 service firebase.storage {
   match /b/{bucket}/o {
     match /{allPaths=**} {
-      allow read, write: if request.auth!=null || $NUMBER == $NUMBER;
+      allow read, write: if request.auth!=null && $NUMBER == $NUMBER;
     }
   }
 }

--- a/scripts/storage-deploy-tests/run.sh
+++ b/scripts/storage-deploy-tests/run.sh
@@ -1,0 +1,54 @@
+#!/usr/bin/env bash
+set -e
+CWD="$(pwd)"
+
+source scripts/set-default-credentials.sh
+
+TARGET_FILE="${COMMIT_SHA}-${CI_JOB_ID}.txt"
+
+echo "Running in ${CWD}"
+echo "Running with node: $(which node)"
+echo "Running with npm: $(which npm)"
+echo "Running with Application Creds: ${GOOGLE_APPLICATION_CREDENTIALS}"
+
+echo "Target project: ${FBTOOLS_TARGET_PROJECT}"
+
+echo "Initializing some variables..."
+DATE="$(date)"
+NUMBER="$(date '+%Y%m%d%H%M%S')"
+echo "Variables initalized..."
+
+echo "Creating temp directory..."
+TEMP_DIR="$(mktemp -d)"
+echo "Created temp directory: ${TEMP_DIR}"
+
+echo "Installing firebase-tools..."
+./scripts/npm-link.sh
+echo "Installed firebase-tools: $(which firebase)"
+
+echo "Initializing temp directory..."
+cd "${TEMP_DIR}"
+cat > "firebase.json" <<- EOM
+{
+  "storage": {
+    "rules": "storage.rules"
+  }
+}
+EOM
+cat > "storage.rules" <<- EOM
+rules_version = '2';
+service firebase.storage {
+  match /b/{bucket}/o {
+    match /{allPaths=**} {
+      allow read, write: if request.auth!=null || $NUMBER == $NUMBER;
+    }
+  }
+}
+EOM
+echo "Initialized temp directory."
+
+echo "Testing storage deployment..."
+firebase deploy --only storage --project "${FBTOOLS_TARGET_PROJECT}"
+RET_CODE="$?"
+test "${RET_CODE}" == "0" || (echo "Expected exit code ${RET_CODE} to equal 0." && false)
+echo "Tested storage deployment."

--- a/src/gcp/storage.ts
+++ b/src/gcp/storage.ts
@@ -1,7 +1,7 @@
 import { Readable } from "stream";
 import * as path from "path";
 
-import api, { storageOrigin } from "../api";
+import api, { appengineOrigin, storageOrigin } from "../api";
 import { Client } from "../apiv2";
 import { logger } from "../logger";
 import { FirebaseError } from "../error";
@@ -128,11 +128,10 @@ interface StorageServiceAccountResponse {
   kind: string;
 }
 
-const storageAPIClient = new Client({ urlPrefix: storageOrigin, apiVersion: "v1" });
-
 export async function getDefaultBucket(projectId?: string): Promise<string> {
   try {
-    const resp = await storageAPIClient.get<{ defaultBucket: string }>(`/apps/${projectId}`);
+    const appengineClient = new Client({ urlPrefix: appengineOrigin, apiVersion: "v1" });
+    const resp = await appengineClient.get<{ defaultBucket: string }>(`/apps/${projectId}`);
     if (resp.body.defaultBucket === "undefined") {
       logger.debug("Default storage bucket is undefined.");
       throw new FirebaseError(
@@ -183,6 +182,7 @@ export async function uploadObject(
   if (path.extname(source.file) !== ".zip") {
     throw new FirebaseError(`Expected a file name ending in .zip, got ${source.file}`);
   }
+  const storageAPIClient = new Client({ urlPrefix: storageOrigin, apiVersion: "v1" });
   const location = `/${bucketName}/${path.basename(source.file)}`;
   const res = await storageAPIClient.request({
     method: "PUT",

